src/docs/sphinx/dev_guide/gpu_porting.rst

```diff
@@ -0,0 +1,302 @@
+`Umpire Tutorial`_ - 1st two sections cover Allocators and Resources
+
+
+.. _gpu-kernels-label:
+
+===============
+Kernels
+===============
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+axom::for_all
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+``axom::for_all`` can be found in
+`axom/core/execution/for_all.hpp <https://github.com/LLNL/axom/blob/develop/src/axom/core/execution/for_all.hpp>`_
+
+``axom::for_all`` is a wrapper around `RAJA forall`_, a simple for loop.
+
+This is used the most in Axom, from unit tests to example drivers for loops
+on device::
+
+  template <typename ExecSpace, typename KernelType>
+  void axom::for_all(const IndexType& N, KernelType&& kernel)
+
+  template <typename ExecSpace, typename KernelType>
+  void axom::for_all(const IndexType& begin, const IndexType& end, KernelType&& kernel)
+
+.. note::
+
+  When Axom is built without RAJA, ``axom::for_all`` becomes a for loop on
+  host (CPU).
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+RAJA::kernel
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+``RAJA::kernel`` is for more complex and/or nested loops.
+
+This is used infrequently, mainly seen only in a `few unit tests`_.
+
+Your general go-to will be ``axom::for_all``.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Useful Links
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+`RAJA Loops`_ - Covers for_all, kernel, and other loops
+
+
+.. _gpu-execution-spaces-label:
+
+================================
+Execution Spaces & Policies
+================================
+
+Axom's execution spaces can be found in
+`axom/core/execution/execution_space.hpp <https://github.com/LLNL/axom/blob/develop/src/axom/core/execution/execution_space.hpp>`_
+
+Axom's execution spaces are derived from an execution_space traits class,
+containing RAJA execution policies and default Umpire memory allocators
+associated with each space.
+
+Axom currently supports `4 execution spaces`_:
+
+* ``SEQ_EXEC`` - Sequential execution policies on host
+* ``OMP_EXEC`` -  OpenMP execution policies on host
+* ``CUDA_EXEC`` - CUDA execution policies in Unified Memory (host + device)
+* ``HIP_EXEC`` - HIP execution policies in Unified Memory (host + device)
+
+Additionally, ``HIP_EXEC`` and ``CUDA_EXEC`` are templated by the
+number of threads and SYNCHRONOUS or ASYNC execution.
+
+.. literalinclude:: ../../../axom/core/execution/internal/cuda_exec.hpp
+   :start-after:  _cuda_exec_start
+   :end-before:  _cuda_exec_end
+   :language: C++
+
+Each execution space (``axom::execution_space<ExecSpace>``) provides:
+
+* Axom policies that are type aliases of RAJA policies to be used with kernels,
+  RAJA types, and RAJA operations
+
+  * ``loop_policy`` - For `RAJA scans`_ and other operations; ``axom::for_all``
+    uses the loop_policy from the templated execution space.
+  * ``reduce_policy`` - For  `RAJA reduction types`_
+  * ``atomic_policy`` - For `RAJA atomic operations`_
+  * ``sync_policy`` - For Axom's `synchronize`_ function, which is a wrapper
+    around ``RAJA::synchronize()``
```

Comment:
A short description of what reduce, atomic, and sync is would be helpful.

Reply:
Also a couple of simple examples of usage would be very helpful.